### ISSUE TITLE
Niche Incorrect Handling Of Nested Async Functions

### ISSUE THREAD

**Issue by zevbo** [python/mypy#13717]
**Bug Report**

If an outer function has a `TypeVar` in any of its arguments, where the `TypeVar` is defined with a list of possiblities, mypy will throw an incorrect type error for the return value of any nested async function. Specifically, it will say that nested function should be returning a coroutine, when obviously it shouldn't.

**To Reproduce**

This is a minimal code snippet to reproduce the bug:
```
 0: from typing import Generic, TypeVar
 1: 
 2: class Type1:
 3:     pass
 4: class Type2:
 5:     pass
 6: 
 7: T = TypeVar("T", Type1, Type2)
 8: 
 9: class A(Generic[T]):
10:   pass
11:
12: async def f(x: A[T]) -> None:
13:    async def f2() -> None:
14:       return None
```

**Expected Behavior**

This should compile fine with mypy with no errors.

**Actual Behavior**

However, on the final line, Mypy says:
`Incompatible return value type (got "None", expected "Coroutine[Any, Any, None]")`

If you make any of the following changes, the error disappears:
- Remove the generic from line 12 so that it looks like `async def f(x: A[T]) -> None:`
- Make `T` defined by a `Union` so that line 7 looks like `T = TypeVar("T", bound=Union[Type1, Type2])`
- Give `A` an explicit type so that lline 12 looks like `async def f(x: A[Type1]) -> None:`

**Your Environment**

- Mypy version used: 0.812
- Mypy command-line flags: no extra flags
- Python version used: 3.8.10
- Operating system and version: Ubuntu

**Comment by AlexWaygood** [comment]
Mypy 0.812 is a very old version of mypy, but this does also reproduce on 0.971: https://mypy-play.net/?mypy=latest&python=3.10&gist=baf0bdea5c503781ec1ba8ee6858eed3

**Comment by jonathanberthias** [comment]
Here is a somewhat simpler example that has the same issue:
```python
from typing import TypeVar

T = TypeVar("T", int, str)

def f(x: T) -> None:
   async def f2() -> None:
      return None  # error: Incompatible return value type (got "None", expected "Coroutine[Any, Any, None]")
```

**Comment by zevbo** [comment]
I would like to bump this as this is a major issue and it is still present

**Comment by erictraut** [comment]
This bug appears to have been fixed in mypy 1.4.0.
